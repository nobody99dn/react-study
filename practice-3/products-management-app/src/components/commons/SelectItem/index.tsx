// Library
import { FilterOrderOptions, ProductTypes } from '@constants/index';
import React, { memo, useCallback } from 'react';

// Style
import './index.css';

interface SelectItemProps {
  id?: string;
  name: string;
  value?: string | number;
  label: string;
  options: string[];
  disable?: boolean;
  handleSelectChange: (value: string, fieldName: string) => void;
}

const SelectItem: React.FC<SelectItemProps> = ({
  id,
  name,
  value,
  label,
  options,
  disable = false,
  handleSelectChange
}) => {
  const handleChange = useCallback(
    (e: { target: { value: string; name: string } }) => {
      handleSelectChange?.(e.target.value, e.target.name);
    },
    [handleSelectChange]
  );

  return (
    <div className='select-wrapper'>
      <label htmlFor={id}>{label}:</label>
      <select
        className='select'
        name={name}
        id={id}
        value={value}
        disabled={disable}
        onChange={handleChange}
        data-testid='select-option'
      >
        <option value=''>Select</option>
        {options.map((option: string, index: number) => (
          <option key={index} value={option}>
            {option}
          </option>
        ))}
      </select>
    </div>
  );
};

export default memo(SelectItem);
