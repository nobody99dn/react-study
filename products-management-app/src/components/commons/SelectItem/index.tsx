// Library
import { FilterOrderOptions, ProductTypes } from '@constants/types';
import React, { memo, useCallback } from 'react';

// Style
import './index.css';

interface SelectItemProps {
  id?: string,
  name: string,
  value?: string | number;
  label: string;
  options: string[];
  onChange: (value: string, fieldName?: string) => void;
}

const SelectItem: React.FC<SelectItemProps> = ({
  id,
  name,
  value,
  label,
  options,
  onChange
}) => {
  const handleChange = useCallback(
    (e: { target: { value: string, name: string; }; }) => {
      onChange?.(e.target.value, e.target.name);
    },
    [onChange],
  );

  return (
    <div className='select-wrapper'>
      <label htmlFor={id}>{label}:</label>
      <select className='select'
        name={name}
        id={id}
        value={value}
        onChange={handleChange}
      >
        <option value="">Select</option>
        {options.map((option: string, index: number) => (
          <option key={index} value={option}>{option}</option>
        ))}
      </select>
    </div>
  );
};

export default memo(SelectItem);
