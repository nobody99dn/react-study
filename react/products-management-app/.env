VITE_BASE_URL=https://62943cf0a7203b3ed0650c9b.mockapi.io
