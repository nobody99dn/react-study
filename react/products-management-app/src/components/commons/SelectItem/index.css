.select-wrapper {
  padding: 1rem 1rem;
}

.select {
  border-radius: 3px;
  font-size: 16px;
  width: 100%;
  outline: none;
  padding: 0.5rem;
  text-indent: 0.25rem;
  background: var(--light-gray);
  border: 1px solid var(--light-gray);
  margin-top: 0.25rem;
}
