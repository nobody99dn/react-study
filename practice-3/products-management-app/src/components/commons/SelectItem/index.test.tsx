// Libraries
import { render, fireEvent, cleanup } from '@testing-library/react';
import { screen, getByLabelText } from '@testing-library/dom';
import { create } from 'react-test-renderer';
import '@testing-library/jest-dom';

// Components
import SelectItem from './index';

describe('Select Item component', () => {
  let container: HTMLElement;

  const defaultProps = {
    label: 'Select Item Label',
    name: 'select',
    options: ['option 1', 'option 2', 'option 3'],
    onChange: jest.fn()
  };

  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    container.remove();
    jest.clearAllMocks();
    cleanup();
  });

  test('should render correctly', () => {
    const defaultSelectItem = create(<SelectItem {...defaultProps} />).toJSON();
    expect(defaultSelectItem).toMatchSnapshot();

    const selectItemWithValue = create(
      <SelectItem {...defaultProps} value={defaultProps.options[1]} />
    ).toJSON();
    expect(selectItemWithValue).toMatchSnapshot();
  });

  test('should render correct label', () => {
    const label: string = 'Select Component';
    render(<SelectItem {...defaultProps} label={label} />, { container });

    expect(container.querySelector('label')).toHaveTextContent(label);
  });

  test('should be click SelectItem', () => {
    const myMock = jest.fn();

    render(<SelectItem {...defaultProps} onChange={myMock} />, { container });

    const select: HTMLElement = container.getElementsByTagName('select')[0];

    fireEvent.change(select);

    expect(myMock).toHaveBeenCalled();
  });
});
