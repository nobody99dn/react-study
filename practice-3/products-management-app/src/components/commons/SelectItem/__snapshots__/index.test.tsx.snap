// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Item component should render correctly 1`] = `
<div
  className="select-wrapper"
>
  <label>
    Select Item Label
    :
  </label>
  <select
    className="select"
    disabled={false}
    name="select"
    onChange={[Function]}
  >
    <option
      value=""
    >
      Select
    </option>
    <option
      value="option 1"
    >
      option 1
    </option>
    <option
      value="option 2"
    >
      option 2
    </option>
    <option
      value="option 3"
    >
      option 3
    </option>
  </select>
</div>
`;

exports[`Select Item component should render correctly 2`] = `
<div
  className="select-wrapper"
>
  <label>
    Select Item Label
    :
  </label>
  <select
    className="select"
    disabled={false}
    name="select"
    onChange={[Function]}
    value="option 2"
  >
    <option
      value=""
    >
      Select
    </option>
    <option
      value="option 1"
    >
      option 1
    </option>
    <option
      value="option 2"
    >
      option 2
    </option>
    <option
      value="option 3"
    >
      option 3
    </option>
  </select>
</div>
`;
